### ISSUE TITLE
gmtspatial -N fails to indentify the points-in-polygon

### ISSUE THREAD

**Issue by joa-quim** [GenericMappingTools/gmt#7745]
See forum post https://forum.generic-mapping-tools.org/t/gmtspatial-issue/4165/5

**Comment by joa-quim** [comment]
I'm very confused with the code. Why, if not all, starting to look only at second half?
```
if (Ctrl->N.all) { first = 0; last = S->n_rows - 1; np = S->n_rows; } else { first = last = S->n_rows / 2; np = 1; }
```
and lower down
```
if (n < np) continue;	/* Not inside this polygon */
```
which jumps to
```
if (Ctrl->N.mode != 1) {	/* Write out results */
	if (GMT_Write_Data (API, GMT_IS_DATASET, GMT_IS_FILE, GMT_IS_POLY, GMT_WRITE_SET, NULL, Ctrl->Out.file, D) != GMT_NOERROR) {
		Return (API->error);
	}
}
```
that writes the non modified input.

**Comment by PaulWessel** [comment]
Looking now.  Some thoughts:

A bit cryptic code, written by me some 15 years ago (gmt5). I dont know if it has changed - the docs are what they are.  The **-N** option talks about points inside features.  This means lines or polygons.  So it does _not_ expect the input to be a bunch of random points but either line or polygon.  Then, per the doc, if any point in a feature (or all points if **+a** is used) is inside any of the polygons in poly then we write out that feature (i.e., a line or polygon). The modifiers **+**r or **+z** affects the output: **+r** just reports the polygon number that contained a feature and **+z** adds a final column with the (same) polygon ID for each output row.  If we agree that this is what the docs says **gmtspatial -N** shall do then there is clearly a bug or two somewhere.

In the process, perhaps another modifier could be created that means "treat all points in a feature as a separate feature", i.e., internally break up input to single points.

**Comment by joa-quim** [comment]
@RustyCuth, you said that it used to work. Do you remember when?

**Comment by PaulWessel** [comment]
Found the issue with that one.  It wrote the output even if nothing is found.

The next issue is the decision that if **+a** is not used (meaning if we find at least one point inside then we report the entire feature) does something that seems dumb: It picks the middle (i.e., point `n_rows / 2 - 1`) and just checks that.  Of course, a snake line could enter the polygon, go outside so the "middle point" is outside, then in again, and it would not be reported as having a point inside!  So I need to change this to try them all and without **+a** we stop that search when the first point inside is found. It works fine with +a.

Last problem is a multiple segment input and if any of them has a segment inside then the entire file, including segments that are outside, are written. Must have forgotten to "turn off" those segments so they dont get written.

Will add a test once done.

**Comment by joa-quim** [comment]
One important ootion would be to treat the input as a point cloud. Then we could ask whose polygon each point fells in. I have a Julia function just for that.

**Comment by PaulWessel** [comment]
Yes, could be useful, but to avoid too much I will now submit my PR for the fixes first, then think about how that new option or modifier can be handled.

**Comment by PaulWessel** [comment]
This output is just for my testing (the actual test is just text comparisons). All seems to work fine.

![map](https://github.com/GenericMappingTools/gmt/assets/26473567/17fefd81-d1c4-4494-b554-de42c5cb6054)

**Comment by PaulWessel** [comment]
Fixed and merged via #7746.
